lms/djangoapps/bookmarks/views.py

```diff
@@ -0,0 +1,301 @@
+
+            * path: (optional) List. List of dicts containing {"usage_id": "", display_name:""} for the XBlocks
+                from the top of the course tree till the parent of the bookmarked XBlock.
+
+            * created: ISO 8601 String. The timestamp of bookmark's creation.
+
+    """
+    authentication_classes = (OAuth2Authentication, SessionAuthentication)
+    permission_classes = (permissions.IsAuthenticated,)
+
+    paginate_by = 30
+    max_paginate_by = 500
+    paginate_by_param = 'page_size'
+    pagination_serializer_class = PaginationSerializer
+    serializer_class = BookmarkSerializer
+
+    def get_serializer_context(self):
+        """
+        Return the context for the serializer.
+        """
+        context = super(BookmarksView, self).get_serializer_context()
+        if self.request.method == 'POST':
+            return context
+        optional_fields_param = self.request.QUERY_PARAMS.get('fields', [])
+        optional_fields = optional_fields_param.split(',') if optional_fields_param else []
+        context['fields'] = DEFAULT_FIELDS + [field for field in optional_fields if field in OPTIONAL_FIELDS]
+        return context
+
+    def get_queryset(self):
+        course_id = self.request.QUERY_PARAMS.get('course_id', None)
+
+        bookmarks_queryset = Bookmark.objects.filter(user=self.request.user)
+
+        if not course_id:
+            return bookmarks_queryset.order_by('-created')
+        try:
+            course_key = CourseKey.from_string(course_id)
+        except InvalidKeyError:
+            log.error("Invalid course id '{course_id}'")
+            return []
+
+        return bookmarks_queryset.filter(course_key=course_key).order_by('-created')
+
+    def post(self, request):
+        """
+        POST /api/bookmarks/v1/bookmarks/
+        """
+        if not request.DATA:
+            error_message = _("No data provided")
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        usage_id = request.DATA.get('usage_id', None)
+        if not usage_id:
+            error_message = _('No usage id provided')
+            return Response(
+                {
+                    "developer_message": error_message,
+                    "user_message": error_message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+
+        try:
+            usage_key = UsageKey.from_string(usage_id)
+        except InvalidKeyError as exception:
+            log.error(exception.message)
+            message = _(u"Invalid usage id")
+            return Response(
+                {
+                    "developer_message": message,
+                    "user_message": message
+                },
+                status=status.HTTP_400_BAD_REQUEST
+            )
+        else:
```

Comment:
Nit: I find this "else" clause confusing. It is only run if the try doesn't thrown an exception, correct? Since the except has a return in it, why the need to next the code below in an "else"?